cpp/src/arrow/record_batch.cc

```diff
@@ -432,4 +433,43 @@ RecordBatchReader::~RecordBatchReader() {
   ARROW_WARN_NOT_OK(this->Close(), "Implicitly called RecordBatchReader::Close failed");
 }
 
+Result<std::shared_ptr<RecordBatch>> ConcatenateRecordBatches(
+    const RecordBatchVector& batches, MemoryPool* pool) {
+  int64_t length = 0;
+  size_t n = batches.size();
+  if (n == 0) {
+    return Status::Invalid("Must pass at least one recordbatch");
+  }
+  if (n == 1) {
+    return batches[0];
+  }
+  int cols = batches[0]->num_columns();
+  auto schema = batches[0]->schema();
+  std::vector<std::shared_ptr<Array>> columns;
+  if (cols == 0) {
+    // special case: null batch, no data, just length
+    for (size_t i = 0; i < batches.size(); ++i) {
+      length += batches[i]->num_rows();
+    }
+  } else {
+    for (int col = 0; col < cols; ++col) {
+      ArrayVector data;
+      for (size_t i = 0; i < batches.size(); ++i) {
+        auto cur_schema = batches[i]->schema();
+        if (!schema->Equals(cur_schema)) {
+          return Status::Invalid(
+              "RecordBatch index ", i, " schema is ", cur_schema->ToString(),
+              ", did not match index 0 recordbatch schema: ", schema->ToString());
+        }
+        auto column_data = batches[i]->column(col);
+        data.push_back(column_data);
+      }
+      auto array = Concatenate(data, pool).ValueOrDie();
```

Comment:
Instead of aborting, please raise the error. We have a helper macro for this:
```suggestion
      ARROW_ASSIGN_OR_RAISE(auto array, Concatenate(data, pool));
```

Reply:
Indeed, return value handling is missing